cmd/gator/validate/validate.go

```diff
@@ -0,0 +1,200 @@
+	Cmd.Flags().BoolVar(&flagJSON, flagNameJSON, false, "print validation information as json.")
+
+	errLog = log.New(os.Stderr, "", 0)
+	outLog = log.New(os.Stdout, "", 0)
+}
+
+func run(cmd *cobra.Command, args []string) {
+	var unstrucs []*unstructured.Unstructured
+
+	// check if stdin has data
+	stdinfo, err := os.Stdin.Stat()
+	if err != nil {
+		errLog.Fatalf("getting info for stdout: %s", err)
+	}
+
+	// using stdin in combination with flags is not supported
+	if stdinfo.Size() > 0 && len(flagFilenames) > 0 {
+		errLog.Fatalf("stdin cannot be used in combination with %q flag", flagNameFilename)
+	}
+
+	// if no files specified, read from Stdin
+	if stdinfo.Size() > 0 {
+		us, err := readYAMLSource(os.Stdin)
+		if err != nil {
+			errLog.Fatalf("reading from stdin: %s", err)
+		}
+		unstrucs = append(unstrucs, us...)
+	} else if len(flagFilenames) > 0 {
+		// normalize directories by listing their files
+		normalized, err := normalize(flagFilenames)
+		if err != nil {
+			errLog.Fatalf("normalizing: %s", err)
+		}
+
+		for _, filename := range normalized {
+			file, err := os.Open(filename)
+			if err != nil {
+				errLog.Fatalf("opening file %q: %s", filename, err)
+			}
+
+			us, err := readYAMLSource(bufio.NewReader(file))
+			if err != nil {
+				errLog.Fatalf("reading from file %q: %s", filename, err)
+			}
+			file.Close()
+
+			unstrucs = append(unstrucs, us...)
+		}
+	} else {
+		errLog.Fatalf("no input data: must include data via either stdin or the %q flag", flagNameFilename)
+	}
+
+	responses, err := validate.Validate(unstrucs)
+	if err != nil {
+		errLog.Fatalf("auditing objects: %v\n", err)
+	}
+
+	results := responses.Results()
+
+	if flagYAML && flagJSON {
+		errLog.Fatalf("cannot use --%s and --%s in combination", flagNameJSON, flagNameYAML)
+	}
+
+	if flagJSON {
+		b, err := json.MarshalIndent(results, "", "    ")
+		if err != nil {
+			errLog.Fatalf("marshaling validation json results: %s", err)
+		}
+		outLog.Fatal(string(b))
+	}
+
+	if flagYAML {
+		b, err := yaml.Marshal(results)
+		if err != nil {
+			errLog.Fatalf("marshaling validation yaml results: %s", err)
+		}
+		outLog.Fatal(string(b))
+	}
+
+	if len(results) > 0 {
+		for _, result := range results {
+			outLog.Printf("Message: %q", result.Msg)
+		}
+
+		os.Exit(1)
+	}
+}
+
+func normalize(filenames []string) ([]string, error) {
+	var output []string
+
+	for _, filename := range filenames {
+		err := filepath.Walk(filename, func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+
+			// only add files to the normalized output
+			if info.IsDir() {
+				return nil
+			}
+
+			output = append(output, path)
+			return nil
+		})
+		if err != nil {
+			return nil, fmt.Errorf("walking %q: %w", filename, err)
+		}
+	}
+
+	return output, nil
+}
+
+func readYAMLSource(r io.Reader) ([]*unstructured.Unstructured, error) {
+	var objs []*unstructured.Unstructured
+
+	decoder := k8syaml.NewYAMLOrJSONDecoder(r, 1000)
```

Comment:
I'm guessing this handles YAML files where objects are separated by `---`?

Reply:
Yes, exactly.  I wanted to leverage yaml ingestion code from the k8s libraries, where I knew it would be battle-tested.

Reply:
I tested and this also supports JSON

Reply:
Yep. I'll have to go back and modify `gator test` to use this pattern.

Julian: could you make these library methods somewhere in `pkg/gator` so I can reuse this logic later?

Reply:
> Julian: could you make these library methods somewhere in pkg/gator so I can reuse this logic later?

Done.  You'll find it in `pkg/gator/read_constraints.go`